modal-js/src/task_command_router_client.ts

```diff
@@ -0,0 +1,475 @@
+          ),
+        1000, // Retry after 1s since total time is expected to be long.
+        1, // Fixed delay.
+        null, // Retry forever.
+        deadline, // Enforce overall deadline.
+      );
+    } catch (err) {
+      if (err instanceof ClientError && err.code === Status.DEADLINE_EXCEEDED) {
+        throw new Error(`Deadline exceeded while waiting for exec ${execId}`);
+      }
+      throw err;
+    }
+  }
+
+  private async refreshJwt(): Promise<void> {
+    this.jwtRefreshLock = this.jwtRefreshLock.then(async () => {
+      if (this.closed) {
+        return;
+      }
+
+      // If the current JWT expiration is already far enough in the future, don't refresh.
+      if (this.jwtExp !== null && this.jwtExp - Date.now() / 1000 > 30) {
+        // This can happen if multiple concurrent requests to the task command router
+        // get UNAUTHENTICATED errors and all refresh at the same time - one of them
+        // will win and the others will not refresh.
+        this.logger.debug(
+          "Skipping JWT refresh because expiration is far enough in the future",
+          "task_id",
+          this.taskId,
+        );
+        return;
+      }
+
+      const resp = await this.serverClient.taskGetCommandRouterAccess(
+        TaskGetCommandRouterAccessRequest.create({ taskId: this.taskId }),
+      );
+
+      if (resp.url !== this.serverUrl) {
+        throw new Error("Task router URL changed during session");
+      }
+
+      this.jwt = resp.jwt;
+      this.jwtExp = parseJwtExpiration(resp.jwt, this.logger);
+    });
+
+    await this.jwtRefreshLock;
+  }
+
+  private async callWithAuthRetry<T>(func: () => Promise<T>): Promise<T> {
+    try {
+      return await func();
+    } catch (err) {
+      if (err instanceof ClientError && err.code === Status.UNAUTHENTICATED) {
+        await this.refreshJwt();
+        return await func();
+      }
+      throw err;
+    }
+  }
+
+  private async *streamStdio(
+    taskId: string,
+    execId: string,
+    fileDescriptor: TaskExecStdioFileDescriptor,
+    deadline: number | null,
+  ): AsyncGenerator<TaskExecStdioReadResponse> {
+    let offset = 0;
+    let delayMs = 10;
+    const delayFactor = 2;
+    let numRetriesRemaining = 10;
+    let numAuthRetries = 0;
+
+    const retryableStatusCodes = new Set([
+      Status.DEADLINE_EXCEEDED,
+      Status.UNAVAILABLE,
+      Status.CANCELLED,
+      Status.INTERNAL,
+      Status.UNKNOWN,
+    ]);
+
+    while (true) {
+      try {
+        const timeoutMs =
+          deadline !== null ? Math.max(0, deadline - Date.now()) : undefined;
+
+        const request = TaskExecStdioReadRequest.create({
+          taskId,
+          execId,
+          offset,
+          fileDescriptor,
+        });
+
+        const stream = this.stub.taskExecStdioRead(request, {
+          timeoutMs,
+        } as CallOptions & TimeoutOptions);
+
+        try {
+          for await (const item of stream) {
+            numAuthRetries = 0;
```

Comment:
Just changed this to a bool in the python implementation: https://github.com/modal-labs/modal-client/pull/3828/changes

Reply:
Changed here too now to match!